### PR TITLE
Add WCCOM requirements check to /installer REST API endpoint

### PR DESCRIPTION
**Pull request by rcstr**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:
Adds internal REST API endpoint to allow WC.COM to validate user's site has met all the requirements to install purchased products.

- Adds API endpoint
    - `GET /wccom-site/v1/installer/requirements_check` te retreive if requirements are met.
- Request is authenticated using `access_token` and `access_token_secret` that are stored in site option after connecting to WooCommerce.com (via __WooCommerce > Extensions > WooCommerce.com Subscriptions__).

### How to test the changes in this Pull Request:
To test this, your WooCommerce site needs to be connected to WooCommerce.com (via __WooCommerce > Extensions > WooCommerce.com Subscriptions__). Once connected, you will need to generate a signature for the request. For example, if your WooCommerce site is https://woo-site.test, signature for `GET /wccom-site/v1/installer/requirements_check` can be generated with:

```php
require_once( WC_ABSPATH . 'includes/admin/helper/class-wc-helper.php' );

$auth                = WC_Helper_Options::get( 'auth' );
$access_token_secret = $auth['access_token_secret'];
$signature           = hash_hmac(
	'sha256',
	json_encode( [
		'host'        => 'woo-site.test:6320',
		'request_uri' => '/wp-json/wccom-site/v1/installer/requirements_check',
		'method'      => 'GET',
	] ),
	$access_token_secret
);
```
Where:

- `$signature` => `X-Woo-Signature: `
- `$auth['access_token']` => `Authorization: Bearer <TOKEN>`

#### Example Request and Response
```
curl -X GET \
  http://woo-site.test:6320/wp-json/wccom-site/v1/installer/requirements_check \
  -H 'Authorization: Bearer bcdc6118bf1bdaa328960b5e5b12cab3cbcad01c' \
  -H 'X-Woo-Signature: e4b39d23cdcf9b549c588151f3dab7bd2c92d0d0f49442e0346d3af7cc35de96'

{"passed":true,"errors":[]}
```
My local instance has met all the necessary requirements, so `passed` is `true` in this case.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Add WCCOM Site Installer REST API endpoint to validate server requirements

### CODE REVIEW

**Comment by haszari** [comment]
@rcstr I tried to test this and I got a 401 permission error, is there something I'm missing? I used `wp shell` to retrieve the `access_token_secret` and generate the signature for my request/host.

```sh
curl   http://woo-site.test:6322/wp-json/wccom-site/v1/installer/requirements_check   -H 'Authorization: Bearer bbf06c5fbd8db44d24fe0db6debc7ea9ff6874fa'   -H 'X-Woo-Signature: f8919f0d973de9ff37c93e9f52d9e7c139b6a94072beb242614a22174b9e63e8'
{"code":"woocommerce_rest_cannot_install_product","message":"You do not have permission to install plugin or theme","data":{"status":401}}
```

```php
// how I generated the signature, my test site is running on woo-site.test:6322
$auth                = WC_Helper_Options::get( 'auth' ); 
$signature           = hash_hmac( \
  'sha256', \
  json_encode( [ \
    'host'        => 'woo-site.test:6322', \
    'request_uri' => '/wp-json/wccom-site/v1/installer/requirements_check', \
    'method'      => 'GET', \
  ] ), \
  $auth['access_token_secret'] \
);
```

**Comment by haszari** [comment]
Comment on this API – could we just add these checks into the install status `GET /install` API endpoint, and return error info in that endpoint? Is it worth having a separate endpoint for this?

**Review by vedanshujain — COMMENTED** [review]
Tested it out, works as expected. Also tested with WP Cron disabled and got an error message as expected like so:
```
{
    "passed": false,
    "errors": [
        "wp_cron"
    ]
}
```

Couple of comments:
1. Not a blocker for this merge, but can we add Unit tests for these endpoints?

2. Same as [Rua's comment](https://github.com/woocommerce/woocommerce/pull/24690#issuecomment-535759981), looks like we won't be able to use anything in installer controller unless this check succeeds, so might as well build these checks in `check_permission` method of `installer` controller? What do you think?

**Comment by kloon** [comment]
@rcstr you can ignore the testing instructions label, we use that internally for our testing.

**Review by tgglv — APPROVED** [review]
Thanks, @rcstr! 
LGTM and works as described.

File: includes/wccom-site/class-wc-wccom-site-installer-requirements-check.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+		if ( ! self::met_filesystem_requirement() ) {
+			$passed   = false;
+			$errors[] = 'filesystem';
+		}
+
+		return array(
+			'passed' => $passed,
+			'errors' => $errors,
+		);
```
[the remark this answers is not in the document]
Comment (kloon) [reply includes/wccom-site/class-wc-wccom-site-installer-requirements-check.php]: @tgglv using short array syntax is actually against WP Coding Standards. It was proposed recently to adopt it but decided to stay with long syntax https://make.wordpress.org/core/2019/07/12/php-coding-standards-changes/

**Review by kloon — DISMISSED** [review]
Happy with all the changes here, can you just update all the since tags to point to 3.8.0 instead as we will not be doing a 3.7.1 release by the looks of things.

**Comment by rcstr** [comment]
Hi team, requested changes have been addressed:

- added an extra validation to our `POST /installer` endpoint in 803f3e6.

Now when the requirements aren't met, the endpoint will return an error with a `503` status and a message containing the unmet requirements.

Note: Travis is failing, the task failing is PHPCS, seems like PHPCS is not being successfully installed in the VM, did restart the job.

**Comment by rcstr** [comment]
I have updated the status error to `400`, in order to avoid unnecessary alerts to customers sites, thanks @haszari for the suggestion.

File: includes/wccom-site/rest-api/endpoints/class-wc-rest-wccom-site-installer-controller.php [hunk 2]
```diff
@@ -95,6 +95,11 @@ public function get_install_state( $request ) {
 	 * @return bool|WP_Error
 	 */
 	public function install( $request ) {
+		$requirements_met = WC_WCCOM_Site_Installer_Requirements_Check::met_requirements();
```
Comment (haszari) [inline includes/wccom-site/rest-api/endpoints/class-wc-rest-wccom-site-installer-controller.php]: Is this only on the `POST` request right now?

Can we add it to the `GET` request also, so there's a lightweight way to check? This allows us to use the `GET` api as a preflight check before allowing the flow (as we originally hoped). I'd suggest we just add this check to all API methods, unless there's a major issue with that (e.g. performance).

**Review by haszari — COMMENTED** [review]
I like the new approach of incorporating this into the existing API as a preflight check.

However, can we add this to more endpoints? In particular the `GET` endpoint, so there's a quick & easy way for WCCOM client to check, so it can "fail fast".

File: includes/wccom-site/rest-api/endpoints/class-wc-rest-wccom-site-installer-controller.php (hunk 2 above)
Comment (rcstr) [reply includes/wccom-site/rest-api/endpoints/class-wc-rest-wccom-site-installer-controller.php]: @haszari  - I have added the validation to `GETm /installer` endpoint in 4fabaed

Comment (haszari) [reply includes/wccom-site/rest-api/endpoints/class-wc-rest-wccom-site-installer-controller.php]: Tested this today, and things are working well :)

- With cron enabled, I was able to auto-install.
- With cron disabled (added `define( 'DISABLE_WP_CRON', true );` to `wp-config.php`), auto install failed after the first site-api call.

**Review by haszari — APPROVED** [review]

**Pull request merged**